Ingest structured/json logs from blobber & make them available via the API

camd opened the pull request:
This PR addresses Bug 1113873
This is the minimum viable product for generating a summary artifact for structured logs.  

1. Adds ``mozlog`` library for parsing structured logs
2. Renames the "Structured Log" artifact to ``text_log_summary`` to remove confusion with the **actual** structured log
3. Updates the ``logslice`` endpoint so that it can slice any log by name, not just the old style raw logs.
4. Abstracts out the creation of the log parse artifact from posting the artifact, so we can reuse that code with the structured log parser
5. Creates a log parser for the structured logs
6. Adds a test for the structured log parser
7. The final commit enables the queue processing for the structured log task so it can easily be turned on and off as needed.

We're deferring the UI for viewing the parsed log this quarter.  So this is mostly as far as we'll get.  (A small UI-side tweak is necessary so that it doesn't try to load this new artifact into the old log parser).  But this opens the door to a Structured log UI being implemented.  The branch is partially complete here: https://github.com/mozilla/treeherder-ui/compare/new-structured-log-viewer

The UI companion PR to this is here:
https://github.com/mozilla/treeherder-ui/pull/399

edmorley commented on treeherder/webapp/api/logslice.py:
```diff
@@ -38,8 +39,8 @@ def list(self, request, project, jm):
         job_id = request.QUERY_PARAMS.get("job_id")
-
-        log = jm.get_log_references(job_id)
+        log_name = request.QUERY_PARAMS.get("name", "builds-4h")
```
I know the 'name' is set to "build-4h" in the logs_url when we ingest it currently (https://github.com/mozilla/treeherder-service/blob/e9780bf1a0eb71fc5ea2f9e5fb847dcf72a4aed0/treeherder/etl/buildapi.py#L172) - however I think this isn't a great name, and as such we should use say "buildbot_text_log" or "text_log" (which will match the new "text_log_summary" name of the artifact that's derived from it).

Normally I'd say this was fodder for another bug, but once we land this (and the corresponding UI part), then we have even more pieces to change when we do change the name. How about having the REST API use the "new" name, and then converting it here to "build-4h" temporarily, until a later bug makes the change to what we put in the DB? (We'll need to have a transition period where the API looks for both names, but that should be simple enough in this future bug).

camd replied:
I'm going to just make the api take the potential new name.  I created https://bugzilla.mozilla.org/show_bug.cgi?id=1139517 to make it store the new name.  There is a lot of test code to modify that I would rather not muddy with this PR.  But great point!  :)

edmorley replied:
Yup that's fine - and what I meant above (though not sure if that was clear) :-)

edmorley commented on treeherder/webapp/api/logslice.py:
```diff
@@ -78,7 +84,10 @@ def list(self, request, project, jm):
 
-                    lines.append({"text": line, "index": i})
+                    if format == 'json':
+                        lines.append({"data": json.loads(line), "index": i})
```
I'm wondering whether it makes sense for the caller of the API to specify the format like this? Or more: it's common for an API to allow switching between a json and text response - but normally this is just the _response_ rather than the loading of the data in the first place? Should we add a 'format' field to the job_log_url table? Or should we just hardcode some of the names here (ie build-4h is always format text), or is it fine for people to ask for the wrong format and we should just catch the exception and display an appropriate error message? (Currently the exception is unhandled if someone requests format=json on a buildbot text log). 

Lots of questions & I don't really know the best answer either - let's talk on a call soon. (And tbh we can always tweak this post-landing; so long as we catch the exception).

camd replied:
Another great point.  Making a change to basically default to text, unless the ``log_name`` is ``mozlog_json`` in which case it will parse the json.  You're right that a param like ``format`` usually has to do with the return value, not the data value.

edmorley commented on treeherder/webapp/api/logslice.py:
```diff
@@ -58,9 +59,14 @@ def list(self, request, project, jm):
+        log = [log for log in logs if log["name"] == log_name]
+
+        if len(log):
+            log = log[0]
```
The 'log' variable initially can actually contain multiple logs, so 'logs' might be a better name. That said, so we ever expect more than one log of each type? I wonder if in a future bug we should enforce that the job_logs_url table only has one entry of each type - or else add support for multiple logs here. Anyway I digress :-)

Seeing as we want to both filter the list of logs and also only return the first matching item, we can use Python's next() for improved perf and fewer lines. I also think it would be clearer to raise the ResourceNotFoundException early. eg:

```python
logs = jm.get_log_references(job_id)
# Get the first log found that has the desired log_name.
log = next([log for log in logs if log["name"] == log_name], None)
if not log:
    raise ResourceNotFoundException("job_artifact {0} not found".format(job_id))

try:
    url = log.get("url")
    ...
finally:
    ...
```

Or:
```python
logs = jm.get_log_references(job_id)
# Get the first log found that has the desired log_name.
try:
    log = next([log for log in logs if log["name"] == log_name])
except StopIteration:
    raise ResourceNotFoundException("job_artifact {0} not found".format(job_id))

try:
    url = log.get("url")
    ...
finally:
    ...
```

camd replied:
Yet another great point!  That second one is the more "pythonic"  I like it.  Thanks for the tip!  I hadn't used ``next()`` before, tbh.

edmorley commented on treeherder/log_parser/utils.py:
```diff
@@ -137,7 +145,14 @@ def get_mozharness_substring(line):
+
+
 def extract_log_artifacts(log_url, job_guid, check_errors):
+    """Generate a summary artifact for the raw (unstructured) log."""
```
Should we use 'text' instead of 'unstructured'?
Just wondering if we should use text vs json when describing them rather than unstructured vs structured, given discussions in the past about "structured logs" being quite a nebulous term. (Yey naming is hard hehe).

edmorley commented on treeherder/log_parser/utils.py:
```diff
@@ -209,3 +224,91 @@ def extract_log_artifacts(log_url, job_guid, check_errors):
+
+def extract_struct_log_artifacts(log_url, job_guid, check_errors):
+    """ Generate a summary artifact for the structured log. """
+
```
I don't think we need this newline :-)

edmorley commented on treeherder/log_parser/utils.py:
```diff
@@ -209,3 +224,91 @@ def extract_log_artifacts(log_url, job_guid, check_errors):
+            oauth_secret=credentials.get('consumer_secret', None),
+        )
+        update_endpoint = 'job-log-url/{0}/update_parse_status'.format(
+            job_log_url['id']
```
Don't need to go to the effort of changing it now, but I think we shouldn't worry about sticking to 80 chars - 100 or 120 seems fine to me (in this case it would only be 89 chars). The flake8 config has the line length check disabled.

edmorley commented on treeherder/log_parser/utils.py:
```diff
@@ -209,3 +224,91 @@ def extract_log_artifacts(log_url, job_guid, check_errors):
+                       extract_artifacts_cb,
+                       check_errors=False):
+    """Post a list of artifacts to a job."""
+
```
Or this newline :-)

edmorley commented on treeherder/etl/buildapi.py:
```diff
@@ -174,6 +174,16 @@ def transform(self, data, filter_to_project=None, filter_to_revision=None,
             else:
                 log_reference = []
```
Now we set/append log_reference in two places, I think we should put the ``log_reference = []`` first, followed by two |if 'foo' in prop: ...| which both log_reference.append().

camd replied:
makes sense, will do.

edmorley commented on treeherder/etl/buildapi.py:
```diff
@@ -174,6 +174,16 @@ def transform(self, data, filter_to_project=None, filter_to_revision=None,
+                    if bf.endswith('_raw.log'):
+                        log_reference.append({
+                            'url': url,
+                            'name': 'structured-raw'
```
For the text logs, we've used "text_log_summary" for the artifact name, and above I proposed using "text_log" or "buildbot_text_log" for the name in job_log_urls. Whereas the naming here will be "structured_log_raw" and "structured_log_summary". I'd like us to be consistent and either use "raw" for both, or for neither. Also, "structured" could apply to any structured log type, and we may use something other than json in the future. How about "json_log" or "blobber_json_log" or ...?

camd replied:
agreed.  I went with ``mozlog_json`` and the api will have ``buildbot_text``.  Another bug will cover converting storage of the artifact from ``builds-4h`` to ``buildbot_text``

edmorley commented on treeherder/log_parser/artifactbuilders.pyx:
```diff
@@ -98,3 +108,60 @@ class BuildbotPerformanceDataArtifactBuilder(ArtifactBuilderBase):
+class StructuredLogArtifactBuilder(ArtifactBuilderBase):
+    """
+    Makes the stuctured log ``struct_faults`` artifact
+
```
Unnecessary newline

edmorley commented on treeherder/log_parser/artifactbuilders.pyx:
```diff
@@ -98,3 +108,60 @@ class BuildbotPerformanceDataArtifactBuilder(ArtifactBuilderBase):
+    Makes the stuctured log ``struct_faults`` artifact
+
+    """
+
```
And here

edmorley commented on treeherder/model/derived/jobs.py:
```diff
@@ -1937,15 +1937,24 @@ def _load_log_urls(self, log_placeholders, job_id_lookup,
-                    task['routing_key'] = 'parse_log.failures'
+
+                if log_ref[1] == 'structured-raw':
+                    # don't parse structured logs for passing tests
```
I wonder if some of our tooling will want to use the output from passing tests too? (eg "compare two test runs" to compare a passing and failing run of a certain job type). Though presume there would be quite a perf hit.

camd replied:
Yeah, I opted initially to only parse failed tests.  I figured we could easily add parsing for passing later by taking out the check.  But, right.  The perf hit could be substantial.  We may need to experiment with that when we cross that bridge.

edmorley commented on treeherder/settings/base.py:
```diff
@@ -193,6 +193,7 @@
     # this is used to give priority to some logs, for example when we need to
     # parse a log on demand
     Queue('log_parser_hp', Exchange('default'), routing_key='parse_log.high_priority'),
+    Queue('log_parser_struct', Exchange('default'), routing_key='parse_log.structured'),
```
log_parser_json ?

edmorley commented on treeherder/log_parser/artifactbuilders.pyx:
```diff
@@ -98,3 +108,60 @@ class BuildbotPerformanceDataArtifactBuilder(ArtifactBuilderBase):
+
+            return None
+
+    def parse_log(self):
```
All of the other parsers in this file have their implementations in parsers.pyx.
Should we move some of this there too? Or ...?

camd replied:
This parsing works a little differently.  Because mozlog parses the whole file itself and the other parsers in ``parsers.pyx`` are line-by-line parsers.  We could rename them as ``LineParserXX`` or something like that.

I could dissect the way ``mozlog`` works to make this a more line-by-line parsing, but I didn't think it was a good idea to do that.

edmorley commented:
Thank you for splitting this up so nicely & for awesome commit messages etc - made it nice and easy to review :-)

Few misc things:
* there is no cap on the max line length, nor a cap on the total fault lines recorded, unlike the text log parsers (a la bug 1078415) - which risks perf issues in the UI + increased DB usage.
* StructuredLogArtifactBuilder isn't referenced at all in  artifactbuildercollection.pyx but yet the other ArtifactBuilders are. I don't really follow the file or class naming in the log_parser directory in general, so this may just be me misunderstanding (I think it could do with some names changes in the future to be clearer) - but not sure whether we're still missing something?
* This raises the flake8 warning count from 0 to 2 - please could you run it locally and fix the warnings? :-)
[~/src/treeherder-service]$ flake8
./tests/log_parser/test_tasks.py:37:1: E302 expected 2 blank lines, found 1
./treeherder/log_parser/tasks.py:55:1: W391 blank line at end of file

Overall this PR looks awesome though - nice work! :-)

wlach commented on treeherder/log_parser/artifactbuilders.pyx:
```diff
@@ -98,3 +108,60 @@ class BuildbotPerformanceDataArtifactBuilder(ArtifactBuilderBase):
+        """
+        handler = self.FaultSummaryHandler()
+
+        fhandle = urllib2.urlopen(
```
It would be better to use 'requests' than urllib2, as it allows connection pooling. See the note here: https://pypi.python.org/pypi/urllib3

Also, this is a total nit, but I find the "fhandle" variable naming pattern overly verbose. Just 'f' gets the point across fine IMO. :)

camd commented:
Awesome, thanks!  I'm actually quite enjoying using SourceTree to break things up.  It's helps me to feel less scattered on a large PR like this, too.  It helps a ton with reviewing your changes as well.  I really like that workflow.

* I added a cap (much like yours, but 200 because it gets warnings and there can be a lot)
* In this case it's no used as a collection of artifact builders, just builds a single artifact from the log.  It could be part of a collection that would build more than one artifact, if we wanted it parsed in several ways.  We may choose to do that at some point.
* fixed the flake8 warnings.  Should be good now.

edmorley commented on treeherder/webapp/api/logslice.py:
```diff
@@ -58,35 +59,45 @@ def list(self, request, project, jm):
+        logs = jm.get_log_references(job_id)
 
+        try:
+            log = next(iter([log for log in logs if log["name"] == log_name]))
```
``iter([a for a in b])`` is equivalent to ``a for a in b``, so this can be written as:
``log = next(log for log in logs if log["name"] == log_name)``

camd replied:
oops, right you are.  fixing now.

edmorley commented on treeherder/log_parser/artifactbuilders.pyx:
```diff
@@ -86,7 +96,7 @@ class BuildbotLogViewArtifactBuilder(ArtifactBuilderBase):
             StepParser(check_errors=check_errors)
         ]
-        self.name = "Structured Log"
+        self.name = "text_log_summary"
```
I've just thought - we'll need to (a) change the UI to reference this new artifact name (https://github.com/mozilla/treeherder-ui/search?q=Structured+Log) and (b) add a temporary conversion in the service, so queries look for both names when trying to retrieve the artefacts.

camd replied:
Yeah, I've addressed that in this PR now: https://github.com/mozilla/treeherder-ui/pull/399 (added it last night).  Adding the handling for both names now in ``webapp/api/artifact.py`` - ``ArtifactViewSet.list``

edmorley commented on treeherder/log_parser/artifactbuilders.pyx:
```diff
@@ -98,3 +108,64 @@ class BuildbotPerformanceDataArtifactBuilder(ArtifactBuilderBase):
+
+        def is_fault(self, message):
+            try:
+                retval = any(["level" in message and
```
Perhaps s/retval/result/ or even just returning on the any() without assigning first?
(To me, retval implies the exit code from launching a process)

camd replied:
true true.  I should just return it.  fixed

edmorley commented on treeherder/webapp/api/logslice.py:
```diff
@@ -58,35 +59,45 @@ def list(self, request, project, jm):
-                    gz_file.close()
-
-        else:
+        except StopIteration:
```
This try-except should only be around the next() line.

edmorley commented on treeherder/webapp/api/artifact.py:
```diff
@@ -31,7 +31,16 @@ def list(self, request, project, jm):
+        qparams = request.QUERY_PARAMS.copy()
+        name = qparams.get('name', None)
+        if name and name == 'text_log_summary':
+            qparams['name__in'] = 'text_log_summary,Structured Log'
```
Awesome :-)

edmorley commented:
The Travis run is failing at the moment on flake8 errors
